libcudacxx/test/libcudacxx/cuda/ptx/ptx.mbarrier.arrive.compile.pass.cpp

```diff
@@ -14,57 +14,142 @@
 #include <cuda/ptx>
 #include <cuda/std/utility>
 
-#include "concurrent_agents.h"
-#include "cuda_space_selector.h"
-#include "test_macros.h"
+/*
+ * We use a special strategy to force the generation of the PTX. This is mainly
+ * a fight against dead-code-elimination in the NVVM layer.
+ *
+ * The reason we need this strategy is because certain older versions of ptxas
+ * segfault when a non-sensical sequence of PTX is generated. So instead, we try
+ * to force the instantiation and compilation to PTX of all the overloads of the
+ * PTX wrapping functions.
+ *
+ * We do this by writing a function pointer of each overload to the `__device__`
+ * variable `fn_ptr`. Now, because weak stores from a single thread may be
+ * elided, we also wrap the store in an if branch that cannot be removed.
+ *
+ * To prevent dead-code-elimination of the if branch, we use
+ * `non_eliminated_false`, which uses inline assembly to hide the fact that is
+ * always false from NVVM.
+ *
+ * So this is how we ensure that none of the function pointer stores are elided.
+ * Because `fn_ptr` is possibly visible outside this translation unit, the
+ * compiler must compile all the functions which are stored.
+ *
+ */
 
-template <typename ... _Ty>
-__device__ inline bool __unused(_Ty...) { return true; }
+__device__ void * fn_ptr = nullptr;
 
-__global__ void test_compilation() {
-  using cuda::ptx::sem_release;
-  using cuda::ptx::space_cluster;
-  using cuda::ptx::space_shared;
-  using cuda::ptx::scope_cluster;
-  using cuda::ptx::scope_cta;
-
-  __shared__ uint64_t bar;
-  bar = 1;
-  uint64_t state = 1;
+__device__ bool non_eliminated_false(void){
```

Comment:
Suggestion: Could we rename this to:
```suggestion
__device__ bool ensure_not_eliminated() {
```

Reply:
Your comment convinced me to make a bigger change. I have removed the `non_eliminated_false` function. Instead, a function pointer of each overload is written to an incrementing pointer. This removes the need for the `non_eliminated_false` function and removes lots of `if` blocks in the test code.